datastore/enrichment.go

```diff
@@ -8,13 +8,22 @@ import (
 	"github.com/quay/claircore/libvuln/driver"
 )
 
+// EnrichmentIter is a function for iterating on enrichment records.
+//
+// It accepts a callback function 'yield' to handle each enrichment record. If the callback
+// returns an error, the iteration is halted, and the error is propagated to indicate cancellation.
+type EnrichmentIter func(yield func(enricher *driver.EnrichmentRecord) error) error
+
```

Comment:
```suggestion
// Iter ...
type Iter[T any] func(yield func(T) bool) error

// EnrichmentIter is a function for iterating on enrichment records.
//
// It accepts a callback function 'yield' to handle each enrichment record. If the callback
// returns an error, the iteration is halted, and the error is propagated to indicate cancellation.
type EnrichmentIter Iter[*driver.EnrichmentRecord]

```

I think specifying a generic type and then additional type definitions would help with having long-ass type definitions.

I'm waffling on changing up the "standard" signature, where the iterator function has no return (meaning the iterator function needs to close over an `error`).

Reply:
@hdonnay:

> I'm waffling on changing up the "standard" signature, where the iterator function has no return (meaning the iterator function needs to close over an error).

Returning an error from the `yield` closure simplifies error handling for the iterator consumer. The closure doesn't have to set outer scope variables to stop the iteration. Secondly, but irrelevant to this change, it could allow the iterator to handle the error within a single iteration (like ignoring or retrying, for example).

Returning an error in the iterator is desirable because it would be complicated to communicate to the iterator consumer that the iteration stopped successfully or due to an error (and actions are necessary, such as rolling back).

```
var err error
vulnIter(func(vuln *claircore.Vulnerability) bool {
	err = writeVuln(vuln)
	return err != nil
}
if err != nil {
	writeVulnRollback()
	return fmt.Errorf("failed to write vuln: %w", err)
}

// Cannot tell if we need to rollback if the iteration stopped due
// an error on the iteration steps, although we are perfectly healthy
// up until now.
```

Reply:
With errors:

```
err := vulnIter(func(vuln *claircore.Vulnerability) bool {
	return writeVuln(vuln)
}
if err != nil {
	// This can be due to both our closure's failure or the iterator's failure to iterate fully.
	writeVulnRollback()
	return fmt.Errorf("failed to write vuln: %w", err)
}

Reply:
> I think specifying a generic type and then additional type definitions would help with having long-ass type definitions.

+1